### PR TITLE
#2592 DynaFlow: Add HVDC to SVC + Refactoring

### PR DESCRIPTION
**Pull request by cossartque**
**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [ ] unit tests and non-regression tests were added (new model, new feature and bug fix)
- [ ] main documentation was updated (update of input/output file, 3rd party, model, repository organization, solver)
- [ ] example documentations were updated (new example in examples folder)
- [ ] the corresponding milestone was added in the ticket and in this PR
- [ ] if this PR modifies the parameters or inputs/outputs of a model/solver: the corresponding xsl was added in util/xsl and platform DB were updated
- [ ] if this PR modifies a dictionary: the corresponding french dictionary was updated

### CODE REVIEW

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo [hunk 1]
```diff
@@ -54,13 +53,196 @@ package BaseClasses
...
+
+    parameter Boolean UseLambda1 = not(Lambda1Pu == 0) "If true, the voltage regulation follows the law U1RefPu = U1Pu + Lambda1Pu * QInj1Pu at terminal 1";
+    parameter Types.PerUnit Lambda1Pu "Parameter Lambda of the voltage regulation law U1RefPu = U1Pu + Lambda1Pu * QInj1Pu, in pu (base UNom, SnRef) at terminal 1";
+    parameter Types.ReactivePower Q1Nom "Nominal reactive power in Mvar at terminal 1";
+
+    input Types.VoltageModulePu U1RefPu(start = U1Ref0Pu) "Voltage regulation set point in pu (base UNom) at terminal 1";
+
+    parameter Types.VoltageModulePu U1Ref0Pu "Start value of the voltage regulation set point in pu (base UNom) at terminal 1";
+
+    Types.ReactivePowerPu QInj1PuQNom "Reactive power at terminal 1 in pu (base Q1Nom) (generator convention)";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo:143]: This is no initial variable, so this line has to be moved between U1RefPu and U1Ref0Pu.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo [hunk 2]
```diff
@@ -54,13 +53,196 @@ package BaseClasses
...
+
+    parameter Boolean UseLambda2 = not(Lambda2Pu == 0) "If true, the voltage regulation follows the law U2RefPu = U2Pu + Lambda2Pu * QInj2Pu at terminal 2";
+    parameter Types.PerUnit Lambda2Pu "Parameter Lambda of the voltage regulation law U2RefPu = U2Pu + Lambda2Pu * QInj2Pu, in pu (base UNom, SnRef) at terminal 2";
+    parameter Types.ReactivePower Q2Nom "Nominal reactive power in Mvar at terminal 2";
+
+    input Types.VoltageModulePu U2RefPu(start = U2Ref0Pu) "Voltage regulation set point in pu (base UNom) at terminal 2";
+
+    parameter Types.VoltageModulePu U2Ref0Pu "Start value of the voltage regulation set point in pu (base UNom) at terminal 2";
+
+    Types.ReactivePowerPu QInj2PuQNom "Reactive power at terminal 2 in pu (base Q2Nom) (generator convention)";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo:159]: This is no initial variable, so this line has to be moved between U2RefPu and U2Ref0Pu.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo [hunk 3]
```diff
@@ -54,13 +53,196 @@ package BaseClasses
...
+      Timeline.logEvent1(TimelineKeys.HVDC1BackRegulation);
+    end when;
+
+    annotation(preferredView = "text");
+  end BaseQStatusDangling;
+
+  partial model BaseQStatus "Base dynamic model QStatus"
+    import Dynawo.NonElectrical.Logs.Timeline;
+    import Dynawo.NonElectrical.Logs.TimelineKeys;
+    extends BaseClasses.BaseQStatusDangling;
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo:198]: Blank line between import and extends

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo [hunk 4]
```diff
@@ -129,6 +310,7 @@ package BaseClasses
   end BaseHvdc;
 
   partial model BaseHvdcP "Base dynamic model for HVDC links with a regulation of the active power"
+    import Modelica;
     extends BaseHvdc(P2Pu(start = s20Pu.re), PInj2Pu(start = - s20Pu.re), Q2Pu(start = s20Pu.im), QInj2Pu(start = - s20Pu.im), terminal2(V(re(start = u20Pu.re), im(start = u20Pu.im)), i(re(start = i20Pu.re), im(start = i20Pu.im))));
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo:314]: Blank line between import and extends

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo [hunk 5]
```diff
@@ -140,17 +322,23 @@ package BaseClasses
...
 
     Types.ComplexApparentPowerPu s2Pu(re(start = s20Pu.re), im(start = s20Pu.im)) "Complex apparent power at terminal 2 in pu (base SnRef) (receptor convention)";
+    Types.Angle Theta1(start = UPhase10) "Angle of the voltage at terminal 1 in rad";
+    Types.Angle Theta2(start = UPhase20) "Angle of the voltage at terminal 2 in rad";
     Types.VoltageModulePu U2Pu(start = ComplexMath.'abs'(u20Pu)) "Voltage amplitude at terminal 2 in pu (base UNom)";
 
     parameter Types.ComplexCurrentPu i20Pu "Start value of complex current at terminal 2 in pu (base UNom, SnRef) (receptor convention)";
     parameter Types.ComplexApparentPowerPu s20Pu "Start value of complex apparent power at terminal 2 in pu (base SnRef) (receptor convention)";
     parameter Types.ComplexVoltagePu u20Pu "Start value of complex voltage at terminal 2 in pu (base UNom)";
+  parameter Types.Angle UPhase10 "Start value of voltage angle and filtered voltage angle at terminal 1 in rad";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo:332]: Indentation to increase

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses_INIT.mo [hunk 6]
```diff
@@ -1,20 +1,84 @@
...
 package BaseClasses_INIT
   extends Icons.BasesPackage;
 
+partial model BasePV_INIT "Base initialization model for PV HVDC"
+
+  parameter Types.ActivePowerPu P1RefSetPu "Start value of active power reference at terminal 1 in pu (base SnRef) (receptor convention)";
+  parameter Types.ReactivePower Q1Nom "Nominal reactive power in Mvar at terminal 1";
+
+  Types.VoltageModulePu U1Ref0PuVar "Start value of the voltage regulation set point in pu (base UNom) at terminal 1";
+  Types.ReactivePowerPu QInj10PuQNom "Reactive power at terminal 1 in pu (base Q1Nom) (generator convention)";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses_INIT.mo:24]: Swap the lines (alphabetical order) and move them after U1Ref0Pu (an initial parameter).

**Review by ErwanGuichard49 — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo [hunk 7]
```diff
@@ -54,13 +53,196 @@ package BaseClasses
...
+
+  protected
+    Types.ReactivePowerPu QInj2RawPu "Raw reactive power at terminal 2 in pu (base SnRef) (generator convention)";
+
+    annotation(preferredView = "text");
+  end BasePTanPhi;
+
+  partial model BasePVDangling "Base dynamic model for PV control at terminal 1"
+
+    parameter Boolean UseLambda1 = not(Lambda1Pu == 0) "If true, the voltage regulation follows the law U1RefPu = U1Pu + Lambda1Pu * QInj1Pu at terminal 1";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo:135]: Should it be a final parameter ?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo [hunk 8]
```diff
@@ -164,6 +352,28 @@ package BaseClasses
...
+    annotation(preferredView = "text",
+      Documentation(info = "<html><head></head><body> This HVDC link regulates the active power flowing through itself. The active power reference is given as an input and can be changed during the simulation. This partial model also implements PQ diagrams at each terminal of the HVDC link.</div></body></html>"));
+  end BaseHvdcPDiagramPQ;
+
+  partial model BaseHvdcPFixedReactiveLimits "Base dynamic model for HVDC links with a regulation of the active power and with fixed reactive power limits"
+    extends BaseHvdcP;
+    extends BaseFixedReactiveLimits;
+
+    annotation(preferredView = "text",
+      Documentation(info = "<html><head></head><body> This HVDC link regulates the active power flowing through itself. The active power reference is given as an input and can be changed during the simulation.</div></body></html>"));end BaseHvdcPFixedReactiveLimits;
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo:375]: New line for  _end BaseHvdcPFixedReactiveLimits;_

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo [hunk 9]
```diff
@@ -207,20 +417,12 @@ package BaseClasses
...
-    PInj2Pu = tableQInj2Min.u[1];
-    PInj2Pu = tableQInj2Max.u[1];
+  partial model BaseHvdcPDanglingFixedReactiveLimits "Base dynamic model for HVDC links with a regulation of the active power and with fixed reactive power limits"
+    extends BaseHvdcPDangling;
+    extends BaseFixedReactiveLimitsDangling;
 
     annotation(preferredView = "text",
-      Documentation(info = "<html><head></head><body> This HVDC link regulates the active power flowing through itself. The active power reference is given as an input and can be changed during the simulation. This partial model also implements PQ diagrams at each terminal of the HVDC link.</div></body></html>"));
-  end BaseHvdcPDiagramPQ;
+      Documentation(info = "<html><head></head><body> This HVDC link regulates the active power flowing through itself. The active power reference is given as an input and can be changed during the simulation. The terminal2 is connected to a switched-off bus.</div></body></html>"));end BaseHvdcPDanglingFixedReactiveLimits;
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo:425]: New line for _end BaseHvdcPDanglingFixedReactiveLimits;_

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo [hunk 10]
```diff
@@ -54,13 +53,196 @@ package BaseClasses
...
+
+    Types.ReactivePowerPu QInj1PuQNom "Reactive power at terminal 1 in pu (base Q1Nom) (generator convention)";
+
+    annotation(preferredView = "text");
+  end BasePVDangling;
+
+  partial model BasePV "Base dynamic model for PV control"
+    extends BaseClasses.BasePVDangling;
+
+    parameter Boolean UseLambda2 = not(Lambda2Pu == 0) "If true, the voltage regulation follows the law U2RefPu = U2Pu + Lambda2Pu * QInj2Pu at terminal 2";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses.mo:151]: Should it be a final parameter ?

**Review by rosiereflo — DISMISSED** [review]

**Review by ErwanGuichard49 — DISMISSED** [review]

File: nrt/data/DynaFlow/cases.py [hunk 11]
```diff
@@ -154,6 +154,17 @@
...
+#  Small System with two HVDC links and a secondary voltage control #
+#####################################################################
+
+case_name = "DynaFlow - HVDC links with a secondary voltage control"
+case_description = "Small System with two HVDC links and a secondary voltage control"
+job_file = os.path.join(os.path.dirname(__file__), "HvdcRpcl2", "HvdcRpcl2.jobs")
+
+test_cases.append((case_name, case_description, job_file, 1, standardReturnCodeType, standardReturnCode))
+
+
```
Comment (ErwanGuichard49) [inline nrt/data/DynaFlow/cases.py:167]: Please delete this blank line

**Review by rosiereflo — APPROVED** [review]

**Pull request merged**